presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalExchangeSinkOperator.java

```diff
@@ -0,0 +1,155 @@
+import static java.util.Objects.requireNonNull;
+
+public class LocalExchangeSinkOperator
+        implements Operator
+{
+    public static class LocalExchangeSinkOperatorFactory
+            implements OperatorFactory, LocalPlannerAware
+    {
+        private final int operatorId;
+        private final LocalExchangeSinkFactory sinkFactory;
+        private final PlanNodeId planNodeId;
+        private boolean closed;
+
+        public LocalExchangeSinkOperatorFactory(int operatorId, PlanNodeId planNodeId, LocalExchangeSinkFactory sinkFactory)
+        {
+            this.operatorId = operatorId;
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.sinkFactory = requireNonNull(sinkFactory, "sinkFactory is null");
+        }
+
+        @Override
+        public List<Type> getTypes()
+        {
+            return sinkFactory.getTypes();
+        }
+
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            checkState(!closed, "Factory is already closed");
+            OperatorContext operatorContext = driverContext.addOperatorContext(operatorId, planNodeId, LocalExchangeSinkOperator.class.getSimpleName());
+            return new LocalExchangeSinkOperator(operatorContext, sinkFactory.createSink());
+        }
+
+        @Override
+        public void close()
+        {
+            if (!closed) {
+                closed = true;
+                sinkFactory.close();
+            }
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            return new LocalExchangeSinkOperatorFactory(operatorId, planNodeId, sinkFactory.duplicate());
+        }
+
+        @Override
+        public void localPlannerComplete()
+        {
+            sinkFactory.noMoreSinkFactories();
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private final LocalExchangeSink sink;
+
+    LocalExchangeSinkOperator(OperatorContext operatorContext, LocalExchangeSink sink)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.sink = requireNonNull(sink, "sink is null");
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return sink.getTypes();
+    }
+
+    @Override
+    public void finish()
+    {
+        sink.finish();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return sink.isFinished();
+    }
+
+    @Override
+    public ListenableFuture<?> isBlocked()
+    {
+        ListenableFuture<?> blocked = sink.waitForWriting();
+        if (blocked.isDone()) {
+            return NOT_BLOCKED;
+        }
+        return blocked;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !isFinished() && isBlocked().isDone();
```

Comment:
BTW, I think when we rewrite the execution code next time, we should reevaluate if this is the best design.